File: solutions/ruby/collatz-conjecture/7/collatz_conjecture.rb
class CollatzConjecture

  class InvalidNumberError < ArgumentError
    def initialize message = 'Number must be a positive integer'
      super
    end
  end

  STEP_COUNTER = 0.step
  COLLATZ = ->(number) { number.even? ? number / 2 : number * 3 + 1 }
  private_constant :STEP_COUNTER, :COLLATZ

  def self.steps number
    new(number).steps
  end

  private

  attr_writer :number, :steps

  def initialize number
    number.positive? or raise InvalidNumberError
    self.number = number
    self.steps = calculate_steps
  end

  def calculate_steps
    current = number

    STEP_COUNTER.count do |step|
      return step if current == 1
      current = COLLATZ[current]
    end
  end

  public

  attr_reader :number, :steps

end

File: solutions/cpp/collatz-conjecture/2/collatz_conjecture.cpp
#include "collatz_conjecture.h"
#include <cstdint>
#include <stdexcept>

namespace collatz_conjecture {

unsigned int steps(int n) {
    if (n < 1) {
        throw std::domain_error("value must be positive");
    }

    std::uint64_t value = static_cast<std::uint64_t>(n);
    unsigned int count = 0;

    while (value > 1) {
        if (value % 2 == 0) {
            value /= 2;
        } else {
            value = value * 3 + 1;
        }

        ++count;
    }

    return count;
}

}
